src/transaction_builder.js

```diff
@@ -217,59 +280,119 @@ TransactionBuilder.prototype.__build = function(allowIncomplete) {
 }
 
 TransactionBuilder.prototype.sign = function(index, privKey, redeemScript, hashType) {
-  assert(this.tx.ins.length >= index, 'No input at index: ' + index)
+  assert(index in this.inputs, 'No input at index: ' + index)
   hashType = hashType || Transaction.SIGHASH_ALL
 
-  var prevOutScript = this.prevOutScripts[index]
-  var prevOutType = this.prevOutTypes[index]
+  var input = this.inputs[index]
 
-  var scriptType, hash
-  if (redeemScript) {
-    prevOutScript = prevOutScript || scripts.scriptHashOutput(redeemScript.getHash())
-    prevOutType = prevOutType || 'scripthash'
+  if (input.hashType !== undefined) {
+    assert.equal(input.hashType, hashType, 'Inconsistent hashType')
+  }
 
-    assert.equal(prevOutType, 'scripthash', 'PrevOutScript must be P2SH')
+  // are we already initialized?
+  if (input.initialized) {
+    if (input.prevOutType === 'scripthash') {
+      assert(input.redeemScript, 'PrevOutScript is P2SH, missing redeemScript')
```

Comment:
`sign` has grown very big. I see this function half the time is responsible for preparing input, the other half doing validations along the way. Can we extract the validation code into a separate function and call `validate` before start prep data & signing?